自分の投稿記事一覧(ユーザー側)

Rihitonnnu opened the pull request:
## 今回のPRで行っていること
- ユーザーの投稿記事のCRUD機能を実装するための一連のファイル作成
- 自分が投稿した記事の一覧画面の表示までを実装
- phpstan.neonの修正

 モデルでのリレーション定義の部分で、phpstanで以下のようなエラーが発生したためphpstan.neonファイルを修正しています。
<img width="760" alt="image" src="https://user-images.githubusercontent.com/69156872/199165394-d453f09c-014d-429c-91d7-7c7683a676db.png">

## UI変更点
### 自分の投稿一覧画面
<img width="945" alt="image" src="https://user-images.githubusercontent.com/69156872/199165136-f4009adc-1e91-4fa4-988c-ca316b0a9f5b.png">

Rihitonnnu committed:
ArticleController作成

Rihitonnnu committed:
ArticleControllerのルーティング追加

Rihitonnnu committed:
Articleモデルの作成

Rihitonnnu committed:
ArticleのFactoryとSeeder作成

Rihitonnnu committed:
articleのマイグレーションファイル作成

Rihitonnnu committed:
DatabaseSeederにArticleSeederを追加

Rihitonnnu committed:
Userモデルでの記事のリレーション定義

Rihitonnnu committed:
ArticleControllerのテスト作成

Rihitonnnu committed:
自分が投稿した記事一覧のbladeファイル作成

Rihitonnnu committed:
ナビゲーションバーに投稿一覧画面へ遷移するリンクを追加

Rihitonnnu committed:
モデルのリレーション定義の際に修正

YutakaOkabe commented on app/Models/User.php:
```diff
@@ -45,6 +46,11 @@ class User extends Authenticatable
         'email_verified_at' => 'datetime',
     ];
 
+    public function article(): HasMany
```
hasManyだったら `articles` で定義した方が良さげ

YutakaOkabe commented on database/seeders/ArticleSeeder.php:
```diff
@@ -0,0 +1,38 @@
+            [
+                'user_id' => 1,
+                'title' => 'タイトル',
+                'content' => '日本国民は、正当に選挙された国会における代表者を通じて行動し、われらとわれらの子孫のために、諸国民との協和による成果と、わが国全土にわたつて自由のもたらす恵沢を確保し、政府の行為によつて再び戦争の惨禍が起ることのないやうにすることを決意し、ここに主権が国民に存することを宣言し、この憲法を確定する。そもそも国政は、国民の厳粛な信託によるものであつて、その権威は国民に由来し、その権力は国民の代表者がこれを行使し、その福利は国民がこれを享受する。これは人類普遍の原理であり、この憲法は、かかる原理に基くものである。われらは、これに反する一切の憲法、法令及び詔勅を排除する。日本国民は、恒久の平和を念願し、人間相互の関係を支配する崇高な理想を深く自覚するのであつて、平和を愛する諸国民の公正と信義に信頼して、われらの安全と生存を保持しようと決 ...
```
日本国憲法ベタうちも好きだけどw、のちのちページネーションのテストしたいってなった時とかで大量のデータを作るのに泣くことになるので faker を使いましょう！
Laravelはデフォルトで fakerを入れてくれてるので、「laravel seed faker」とかで使い方調べてみてー！

YutakaOkabe commented on tests/Feature/User/ArticleControllerTest.php:
```diff
@@ -0,0 +1,27 @@
+    /**
+     * @test
+     */
+    public function ログインしていれば自分の投稿一覧を表示する()
```
このテストケースの中で自分のArticleだけが表示されていること（他のユーザーのArticleもダミーで一個作ってみたりして）と、別のテストケースでログインしていない場合にどうなるか、も追加してみてー！

YutakaOkabe reviewed (changes requested):
大体いい感じ〜！いくつか要望書いてるので修正してみてー！

Rihitonnnu committed:
リレーションのメソッド名変更

hasManyなのでメソッドを複数形のarticlesに変更しました

Rihitonnnu committed:
自分の記事のみが表示できているかどうかのテストを追加

Rihitonnnu committed:
投稿一覧表示でログインしていない場合のテストケース追加

Rihitonnnu committed:
記事のデータをfakerを用いて追加するように修正

Rihitonnnu committed:
AdminFactoryのfakerのプロパティ一部変更

Rihitonnnu commented:
@YutakaOkabe 
修正しましたのでレビューお願いします🙇

YutakaOkabe commented on tests/Feature/User/ArticleControllerTest.php:
```diff
@@ -0,0 +1,53 @@
+    /**
+     * @test
+     */
+    public function ログインしていれば自分の投稿一覧を表示する()
```
テストの書き方の好みではあるんだけど、できるだけ以下のことは避けたい
- idを指定したテスト <- DBがリフレッシュできてなかった場合にテストが落ちる可能性がある
- マジックナンバーを使ったテスト（ `$this->user[0]` の0とか）<- コードの視認性が悪い

なので、おれであればこんな感じでここのテストは書くかな！
```
class ArticleControllerTest extends TestCase
{
    protected function setUp(): void
    {
        parent::setUp();
        $this->user = User::factory()->create();
    }

    /**
     * @test
     */
    public function ログインしていれば自分の投稿一覧を表示する()
    {
        $this->actingAs($this->user, 'users');
        $article = Article::factory()->create(['user_id' => $this->user->id]);
        $otherUserArticle = Article::factory()->create(['user_id' => User::factory()->create()->id]);        

        $response = $this->get(route('user.article.index'));

        $response->assertStatus(200);
        $response->assertSeeText($article->title);
        $response->assertDontSeeText($otherUserArticle->title);
    }
}

Rihitonnnu replied:
> idを指定したテスト <- DBがリフレッシュできてなかった場合にテストが落ちる可能性がある

DBがリフレッシュできない場合も考えないといけないんですね...気をつけます！！

> マジックナンバーを使ったテスト（ $this->user[0] の0とか）<- コードの視認性が悪い

確かに具体的な名前がついていない分、一瞬でパッとわかるコードになっていないですね。修正しておきます！

YutakaOkabe commented on tests/Feature/User/ArticleControllerTest.php:
```diff
@@ -0,0 +1,53 @@
+    {
+        parent::setUp();
+        $this->user = User::factory()->count(2)
+            ->state(new Sequence(
+                ['id' => 1],//自分
+                ['id' => 2],//他のユーザー
```
ほーこんな書き方できるのね！勉強になった！

YutakaOkabe commented on tests/Feature/User/ArticleControllerTest.php:
```diff
@@ -0,0 +1,53 @@
+
+        $response = $this->get(route('user.article.index'));
+
+        $getArticle = Article::where('user_id', $this->user[0]->id)->get();
+
+        $this->assertNotEquals(2, $getArticle[0]->user_id);//自分の記事以外が含まれていないかの確認
```
controllerでの処理内容が正しいことをテストしたいので、テストケース側でDBからデータをとってきちゃうのはあんまり意味がないね

YutakaOkabe reviewed (changes requested):
ほとんど良さげ！よく書けてる！
またテストのところで色々いってるので確認してもらえたらー！

Rihitonnnu committed:
ArticleFactoryのuser_id部分を修正

Rihitonnnu committed:
ArticleControllerTestの投稿記事一覧表示のテストケース修正

以下の問題を解決するためにテストケースを修正しました。

idを指定することで落ちる可能性が考えられるテストケースになっている点
マジックナンバーを用いた可読性の低いテスト担っている点

Rihitonnnu commented:
レビューお願いします！

YutakaOkabe commented on app/Http/Controllers/User/ArticleController.php:
```diff
@@ -0,0 +1,25 @@
+
+class ArticleController extends Controller
+{
+    protected Article $article;
```
こいつprivateでいけないかな？
できるだけ変数とメソッドのスコープは狭くした方がコードは読みやすい！

Rihitonnnu replied:
了解です！修正しておきます！

Rihitonnnu replied:
こいつですがprivateにしたところ、クラス内でこのプロパティで使用していないためstanでエラーが発生しました。
なので以下の記述部分は一旦削除しています。

```
protected Article $article;

public function __construct(Article $article)
{
        $this->article = $article;
}
```

YutakaOkabe commented on tests/Feature/User/ArticleControllerTest.php:
```diff
@@ -0,0 +1,44 @@
+    {
+        $this->actingAs($this->user, 'users');
+
+        $sameUserArticle = Article::factory()->create(['user_id' => $this->user->id]); //自分の投稿分
```
sameって何とsameだ？？ってなったので、userArticleでよさそう！
setUpに$this->userがいることで、多分こいつがこのクラスの主体なんだろうなーってわかるので

Rihitonnnu replied:
ここちょっと迷ってました😅でも変数名でよりわかりにくくしてるのはだめですもんね...修正しておきます！

Rihitonnnu replied:
ここちょっと迷ってました😅でも変数名でよりわかりにくくしてるのはだめですもんね...修正しておきます！

YutakaOkabe reviewed (approved):
細かいところコメントしてるけどあとは良さげ！LGTM！
:gangan:

Rihitonnnu committed:
コントローラ内で用いる変数のスコープを変更

Rihitonnnu committed:
自分の記事のデータを格納している変数の名前を変更

Rihitonnnu committed:
使用していないコンストラクタや変数の削除

Rihitonnnu commented:
レビューして頂きありがとうございました🙇‍♀！
